snips_nlu/intent_parser/lookup_intent_parser.py

```diff
@@ -0,0 +1,426 @@
+            self._intents_mapping[intent_name] = intent_id
+
+        return intent_id
+
+    def get_slot_id(self, slot_name):
+        """generate a numeric id for a slot
+
+        Args:
+            slot_name (str): intent name
+
+        Returns:
+            int: numeric id
+
+        """
+        slot_id = self._slots_mapping.get(slot_name)
+        if slot_id is None:
+            slot_id = len(self.slots_names)
+            self.slots_names.append(slot_name)
+            self._slots_mapping[slot_name] = slot_id
+
+        return slot_id
+
+    def _preprocess_text(self, txt):
+        """Replaces stop words and characters that are tokenized out by
+            whitespaces"""
+        tokens = tokenize_light(txt, self.language)
+        cleaned_string = " ".join(
+            [tkn for tkn in tokens if normalize(tkn) not in self.stop_words]
+        )
+        return cleaned_string.lower()
+
+    def generate_io_mapping(self, intents, entity_placeholders):
+        """Generate input-output pairs"""
+        for intent_name, intent in iteritems(intents):
+            intent_id = self.get_intent_id(intent_name)
+            for entry in intent[UTTERANCES]:
+                yield self._build_io_mapping(
+                    intent_id, entry, entity_placeholders
+                )
+
+    def _build_io_mapping(self, intent_id, utterance, entity_placeholders):
+        input_ = []
+        output = [intent_id]
+        slots = []
+        for chunk in utterance[DATA]:
+            if SLOT_NAME in chunk:
+                slot_name = chunk[SLOT_NAME]
+                slot_id = self.get_slot_id(slot_name)
+                entity_name = chunk[ENTITY]
+                placeholder = entity_placeholders[entity_name]
+                input_.append(placeholder.lower())
+                slots.append(slot_id)
+            else:
+                input_.append(chunk[TEXT])
+        output.append(slots)
+
+        key = self._preprocess_text(" ".join(input_))
+
+        return key, output
+
+    def _replace_entities_with_placeholders(self, text, entities):
+        if not entities:
+            return text
+        entities.sort(key=lambda e: e[RES_MATCH_RANGE][START])
+        processed_text = ""
+        current_idx = 0
+        for ent in entities:
+            start = ent[RES_MATCH_RANGE][START]
+            end = ent[RES_MATCH_RANGE][END]
+            processed_text += text[current_idx:start]
+            place_holder = _get_entity_name_placeholder(
+                ent[ENTITY_KIND], self.language
+            )
+            processed_text += place_holder
+            current_idx = end
+        processed_text += text[current_idx:]
+
+        return processed_text
+
+    @check_persisted_path
+    def persist(self, path):
+        """Persists the object at the given path"""
+        path.mkdir()
+        parser_json = json_string(self.to_dict())
+        parser_path = path / "intent_parser.json"
+
+        with parser_path.open(mode="w") as pfile:
```

Comment:
The encoding should be specified here: `parser_path.open(encoding="utf8", mode="w")`